File: Program2/minHeap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;

namespace ConsoleApp1
{
    class MinHeap
    {
        // member variables
        List<HeapNode> m_heapArray = new List<HeapNode>(); // list to store heap nodes
        private int m_size = 0; // size of heap

        /*
         * Empty Constructor
         */
         public MinHeap()
        {
            m_size = 0;
        }

        /*
         * Single Value constructor 
         */
         public MinHeap(HeapNode node)
        {
            m_heapArray.Add(node); // add node to heap array
            m_size = 1; //size is 1
        }

        /*
         * Build heap in linear time Constructor
         */
         public MinHeap(List<HeapNode> heapArray)
        {
            m_heapArray = heapArray.GetRange(0, heapArray.Count); //copy input array into internal array
            m_size = heapArray.Count();
            BuildHeap();
        }

        /*
         * Linear time build of heap using float down on first n//2 nodes
         */
         private void BuildHeap()
        {
            //loop over each n//2 nodes (starting at last leaf node and work to root)
            //and float down each node to proper position in heap
            for (int i = m_size / 2; i >= 0; i--)
            {
                FloatDown(i);
            }
        }

        /*
         * Float Down method to float nodes down to proper location in heap
         */
         private void FloatDown(int index)
        {
            int leftchild = 2 * index + 1; //left child in heap array
            int rightchild = 2 * index + 2; // right child in heap array
            if (index < 0 || m_size <= leftchild) //invalid index
            {
                ;
            }
            else if (m_size <= rightchild) // check if right child exists 
            {
                // no right child so just check if leftchild is less than parent
                if (m_heapArray[index].GetPriority() > m_heapArray[leftchild].GetPriority() )
                {
                    SwapNode(index, leftchild);
                    FloatDown(leftchild);
                }
            }
            else
            {
                // left and right child exists
                if (m_heapArray[index].GetPriority() > m_heapArray[leftchild].GetPriority() ||
                    m_heapArray[index].GetPriority() > m_heapArray[rightchild].GetPriority())
                {
                    int minchild;
                    if (m_heapArray[leftchild].GetPriority()  < m_heapArray[rightchild].GetPriority())
                    {
                        minchild = leftchild;
                    }
                    else
                    {
                        minchild = rightchild;
                    }
                    SwapNode(index, minchild);
                    FloatDown(minchild);
                }
            }
        }

        /*
         * Swap Nodes of heap array
         */
         private void SwapNode(int parent, int child)
        {
            HeapNode temp = m_heapArray[parent];
            m_heapArray[parent] = m_heapArray[child];
            m_heapArray[child] = temp;
        }

        /*
         * Float a node up to its correct position.
         */
         private Boolean FloatUp(int index)
        {
            Boolean swapped = false;
            int parent = index / 2; // parent index of node
            if (parent >= 0 && m_heapArray[index].GetPriority() < m_heapArray[parent].GetPriority())
            {
                SwapNode(parent, index);
                swapped = true;
            }
            return swapped;
        }

        /*
         * Public method for inserting node into heap array
         * If heap array is empty, simply adds the node
         * Else it inserts node at end of heap array and floats the node up to the correct position
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void InsertNode(HeapNode node)
        {
            if (m_size == 0)
            {
                m_heapArray.Add(node);
                m_size += 1;
            }
            else
            {
                m_heapArray.Add(node);
                m_size += 1;
                Boolean swapped = true;
                int index = m_size - 1;
                while (swapped)
                {
                    swapped = FloatUp(index);
                    index = index / 2;
                }
            }
        }

        /*
         * Get root node of heap
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public HeapNode GetMin()
        {
            if (m_size != 0)
            {
                return m_heapArray[0];
            }
            else
            {
                throw new Exception();
            }
        }

        /*
         * Remove the minimum (root) node from heap
         * Swap root with last node, pop the last node and then recursively call float down on root node
         * to reestablish heap property
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public HeapNode RemoveMin()
        {
            HeapNode retNode;
            if (m_size == 0)
            {
                throw new Exception();
            }
            else
            {
                retNode = m_heapArray[0]; // current root
                m_heapArray[0] = m_heapArray[m_size - 1]; // set root node to last node
                m_heapArray.RemoveAt(m_size - 1);
                m_size--;
                FloatDown(0);
            }
            return retNode;
        }

        /*
         * Print Heap
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void PrintHeap()
        {
            if (m_size == 0)
            {
                Console.WriteLine("Heap is empty.");
            }
            for (int i = 0; i < m_size; i++)
            {
                Console.WriteLine(m_heapArray[i]);
            }
        }

        /*
         * Get size of heap
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public int Size()
        {
            return m_size;
        }
    }
}

File: Program1/MinHeap.java
/*
 * A minimum heap to use as a priority queue that is synchronized to allow use in multi-threaded programs
 * Uses an array list as its underlying structure for storing the heap
 */

import java.util.ArrayList;

public class MinHeap {
    ArrayList<HeapNode> m_heapArray = new ArrayList<HeapNode>(); // Array list used for heap
    private int m_size = 0; // size of heap

    /*
     * Empty constructor
     */
    MinHeap() {
    }

    /*
     * constructor for passing in a single heap node
     */
    MinHeap(HeapNode node) {
        m_heapArray.add(node); // add node to heap as root
        m_size = 1; // update size
    }

    /*
     * Array list constructor to build heap in linear time
     */
    MinHeap(ArrayList<HeapNode> heapArray) {
        m_heapArray.addAll(heapArray); // copy input array into internal array
        m_size = heapArray.size(); // update size
        buildHeap();
    }

    /*
     * Linear time build of heap using float down of first n//2 nodes
     */
    private void buildHeap() {
        // loop over each n//2 nodes (starting at last non leaf node and work to root)
        // and float each node down to proper position
        for (int i = Math.floorDiv(m_size, 2); i >= 0; i--) {
            int index = i; // first non-leaf node is at n//2 - 1 for a 0 indexed array
            floatDown(index); // start recursive call of float down for index
        }
    }

    /*
     * Float a node from its position up to the correct position
     */
    private boolean floatUp(int index) {
        int parent = Math.floorDiv(index, 2); // parent of current node
        boolean ret = false;
        if (parent < 0) { // check that current node is not root
            ret = false;
        }
        else {
            // if parent node is greater than current node, swap nodes
            if (m_heapArray.get(parent).getPriority() > m_heapArray.get(index).getPriority()) {
                swapNode(index, parent);
                ret = true;
            }
            else {
                ret = false;
            }
        }
        return ret;
    }

    /*
     * Swap node of parent with child
     */
    private void swapNode(int parent, int child) {
        HeapNode temp = m_heapArray.get(parent);
        m_heapArray.set(parent, m_heapArray.get(child));
        m_heapArray.set(child, temp);
    }

    /*
     * public method call for inserting node
     * if heap is empty, simply adds the node
     * if heap is not empty then it inserts node at the end and floats it up
     */
    synchronized public void insertNode(HeapNode node) {
        if (isEmpty()) {
            m_heapArray.add(node);
            m_size += 1;
        }
        else {
            m_heapArray.add(node);
            m_size += 1;
            boolean swapped = true;
            int index = m_size-1;
            while (swapped == true) {
                swapped = floatUp(index);
                index = Math.floorDiv(index, 2);
            }
        }
    }

    /*
     * Float down for building heap
     * Floats a node down recursively to find the correct position for a node during the build heap process
     */
    private boolean floatDown(int index) {
        int leftchild = 2 * index + 1; // left child in an array heap
        int rightchild = 2 * index + 2; // right child in an array heap
        boolean ret = false;
        if (index < 0) { // invalid index
            ret = false;
        }
        else if (m_size <= leftchild) { // check if left child exists
            ret = false;
        }
        else if (m_size <= rightchild) { // check if right child exists
            // right child does not exist so just check if parent is less then left child
            // if so then swap nodes and recursively call floatdown on child node
            if (m_heapArray.get(index).getPriority() > m_heapArray.get(leftchild).getPriority()) {
                swapNode(index, leftchild);
                floatDown(leftchild);
                ret = true;
            }
            else {
                ret = false;
            }
        }
        else {
            // left and right exist
            // same structure as previous, but now just check if either is less and if so then grab minimum child node
            // recursively call floatdown on min node if swapping
            if (m_heapArray.get(index).getPriority() > m_heapArray.get(leftchild).getPriority()
                    || m_heapArray.get(index).getPriority() > m_heapArray.get(rightchild).getPriority()) {
                int minchild;
                if (m_heapArray.get(leftchild).getPriority() < m_heapArray.get(rightchild).getPriority()) {
                    minchild = leftchild;
                }
                else {
                    minchild = rightchild;
                }
                swapNode(index, minchild);
                floatDown(minchild);
                ret = true;
            }
        }
        return ret;
    }

    /*
     * return the root node without removing the root node
     */
    synchronized public HeapNode getMin() {
        return m_heapArray.get(0);
    }

    /*
     * check if heap is empty
     */
    synchronized public boolean isEmpty() {
        boolean ret = false;
        if (m_size == 0) {
            ret = true;
        }
        else {
            ret = false;
        }
        return ret;
    }

    /*
     * remove the minimum (root) node from the heap
     * Swap root with last node, pop the last node and then recursively call float down on root node to
     * re-establish heap property
     */
    synchronized public HeapNode removeMin() throws MyEmptyHeapException {
        HeapNode retNode;
        if (isEmpty()) { // cant pop root from an empty heap
            throw new MyEmptyHeapException("Heap is empty");
        }
        else {
            retNode = m_heapArray.get(0); // current root
            m_heapArray.set(0, m_heapArray.get(m_size-1)); //set root to last value of heap
            m_heapArray.remove(m_size-1); // remove last value
            m_size -= 1; //decrease size
            floatDown(0); // float root down
        }
        return retNode;
    }

    /*
     * simplistic print of heap for error checking
     */
    synchronized public void printHeap() {
        if (m_size == 0) {
            System.out.println("Heap is empty.");
        }
        for (int i = 0; i < m_size; i++) {
            System.out.println(m_heapArray.get(i));
        }
    }

    /*
     * simple check of heap property for error check
     * checks to make sure every node's child is greater than or equal to itself
     */
    synchronized public boolean checkHeapProperty() {
        boolean ret = true;
        for (int i=0; i < Math.floorDiv(m_size, 2); i++) {
            if (!checkHeapPropertyHelp(i)) {
                ret = false;
                break;
            }
        }
        return ret;
    }

    /*
     * helper function for check heap property to start the recursive call stack
     */
    private boolean checkHeapPropertyHelp(int index) {
        int leftchild = index * 2 + 1;
        int rightchild = index * 2 + 2;
        boolean ret = true;
        if (leftchild >= m_size) {
            ret = true;
        }
        else if (rightchild >= m_size) {
            if (m_heapArray.get(leftchild).getPriority() < m_heapArray.get(index).getPriority()) {
                ret = false;
            }
        }
        else if (m_heapArray.get(leftchild).getPriority() < m_heapArray.get(index).getPriority()
            || m_heapArray.get(rightchild).getPriority() < m_heapArray.get(index).getPriority()) {
            ret = false;
        }
        else ;
        return ret;
    }

    /*
     * return the size of the heap
     */
    synchronized int size() {
        return m_size;
    }
}
